### PR TITLE
Fourier

### PR DESCRIPTION
**Pull request by eggrobin**
Add a continuous Fourier transform on piecewise Poisson series, for use in Chapront’s frequency search to refine the output of the DFT.

Another flying buttress for #2400.

The autoformatter went wild again; kept some of what it did where some code was egregiously poorly formatted (`4* Sqrt(2)`???).

### CODE REVIEW

File: numerics/poisson_series.hpp [hunk 1]
```diff
@@ -299,6 +301,18 @@ class PiecewisePoissonSeries {
   // t must be in the interval [t_min, t_max].
   Value operator()(Instant const& t) const;
 
+  // Returns the Fourier transform of this Poisson series on [t_min, t_max].
+  // When evaluated at a given frequency, the Fourier transform is computed by
+  // Gauss-Legendre quadrature on each subinterval, where the number of points
+  // is chosen assuming that the periods of periodic terms are all large
+  // compared to the subintervals.
+  // If apodization is desired, the |*this| should be multiplied by an
```
Comment (pleroy) [inline numerics/poisson_series.hpp:309]: Remove "the", it reads odd.

File: numerics/poisson_series.hpp [hunk 2]
```diff
@@ -299,6 +301,18 @@ class PiecewisePoissonSeries {
...
+  // apodization function, and |FourierTransform| should be called on the
+  // product.
+  // |*this| must outlive the resulting function.
+  std::function<Complexification<Value>(AngularFrequency const&)>
```
Comment (pleroy) [inline numerics/poisson_series.hpp:313]: I'd favour giving a name to this type.

File: numerics/poisson_series_body.hpp [hunk 3]
```diff
@@ -653,6 +653,33 @@ Value PiecewisePoissonSeries<Value, degree_, Evaluator>::operator()(
   return series_[it - bounds_.cbegin() - 1](t);
 }
 
+template<typename Value,
+         int degree_,
+         template<typename, typename, int>
+         class Evaluator>
```
Comment (pleroy) [inline numerics/poisson_series_body.hpp:659]: Move the end of the previous line (look at the code around you).

File: numerics/poisson_series_body.hpp [hunk 4]
```diff
@@ -653,6 +653,33 @@ Value PiecewisePoissonSeries<Value, degree_, Evaluator>::operator()(
...
+inline std::function<Complexification<Value>(AngularFrequency const&)>
```
Comment (pleroy) [inline numerics/poisson_series_body.hpp:660]: Remove `inline`.

File: numerics/poisson_series_body.hpp [hunk 5]
```diff
@@ -653,6 +653,33 @@ Value PiecewisePoissonSeries<Value, degree_, Evaluator>::operator()(
...
+  // Gaussian quadratures, removing the lifetime requirement on |*this| and
+  // potentially speeding up repeated evaluations of the Fourier transform.
+  return [this](AngularFrequency const& ω) {
+    Interval<Instant> const time_domain{t_min(), t_max()};
+    Instant const t0 = time_domain.midpoint();
+    Primitive<Complexification<Value>, Instant> integral;
+    for (int k = 0; k < series_.size(); ++k) {
+      integral += quadrature::GaussLegendre<std::max(1, (degree_ + 1) / 2)>(
+          [&f = series_[k], t0, ω](
+              Instant const t) -> Complexification<Value> {
```
Comment (pleroy) [inline numerics/poisson_series_body.hpp:672]: Pass `Instant` by reference.  For better or for worse, that's what we do everywhere.

File: numerics/poisson_series.hpp (hunk 2 above)
Comment (pleroy) [reply numerics/poisson_series.hpp:313]: Also, this file should include `functional`.

File: numerics/poisson_series_test.cpp [hunk 6]
```diff
@@ -199,11 +205,63 @@ TEST_F(PoissonSeriesTest, PointwiseInnerProduct) {
...
+             Displacement<World>({-7 * Metre, 11 * Metre, -13 * Metre})}}});
+  // Slice our signal into segments short enough that one-point Gauss-Legendre
+  // (also known as midpoint) does the job.
+  constexpr int segments = 100;
+  PiecewisePoissonSeries<Displacement<World>, 0, HornerEvaluator> f(
+      {t0_, t0_ + π * Second / segments}, signal);
+  for (int k = 1; k < segments; ++k) {
+    f.Append({t0_ + k * π * Second / segments,
+              t0_ + (k + 1) * π * Second / segments},
+             signal);
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:226]: I would call `AtOrigin` to put the origin of `signal` in the middle of the interval, since it would correspond more closely to what we do in real life.

File: numerics/poisson_series_test.cpp [hunk 7]
```diff
@@ -199,11 +205,63 @@ TEST_F(PoissonSeriesTest, PointwiseInnerProduct) {
...
+  }
+  auto const f_fourier_transform = f.FourierTransform();
+  auto const f_power_spectrum =
+      [&f_fourier_transform](AngularFrequency const& ω) {
+        return f_fourier_transform(ω).Norm²();
+      };
+  EXPECT_THAT(Brent(f_power_spectrum, 0.9 * ω, 1.1 * ω, std::greater<>{}),
+              IsNear(4.117_⑴ * Radian / Second));
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:234]: I would write this `RelativeErrorFrom(ω, IsNear(5%))`.  That would give the reader a better sense of the accuracy.

File: numerics/poisson_series_test.cpp [hunk 8]
```diff
@@ -199,11 +205,63 @@ TEST_F(PoissonSeriesTest, PointwiseInnerProduct) {
...
+  // A peak arising from the finite interval.
+  EXPECT_THAT(Brent(f_power_spectrum,
+                    0 * Radian / Second,
+                    2 * Radian / Second,
+                    std::greater<>{}),
+              IsNear(1.209_⑴ * Radian / Second));
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:240]: Could we compute the expected value of this extra peak?

File: numerics/poisson_series_test.cpp [hunk 9]
```diff
@@ -199,11 +205,63 @@ TEST_F(PoissonSeriesTest, PointwiseInnerProduct) {
...
+
+  auto const fw = f * apodization::Hann<HornerEvaluator>(f.t_min(), f.t_max());
+
+  auto const fw_fourier_transform = fw.FourierTransform();
+  auto const fw_power_spectrum =
+      [&fw_fourier_transform](AngularFrequency const& ω) {
+        return fw_fourier_transform(ω).Norm²();
+      };
+  EXPECT_THAT(Brent(fw_power_spectrum, 0.9 * ω, 1.1 * ω, std::greater<>{}),
+              IsNear(+3.979_⑴ * Radian / Second));
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:250]: Same as above.

File: numerics/poisson_series_test.cpp [hunk 10]
```diff
@@ -199,11 +205,63 @@ TEST_F(PoissonSeriesTest, PointwiseInnerProduct) {
...
+  // Hann filters out the interval; this search for a second maximum converges
+  // to its boundary.
+  EXPECT_THAT(Brent(fw_power_spectrum,
+                    0 * Radian / Second,
+                    2 * Radian / Second,
+                    std::greater<>{}),
+              IsNear(1.9999999_⑴ * Radian / Second));
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:257]: Nice.

File: numerics/poisson_series_test.cpp [hunk 11]
```diff
@@ -199,11 +205,63 @@ TEST_F(PoissonSeriesTest, PointwiseInnerProduct) {
...
+}
+
 TEST_F(PoissonSeriesTest, Primitive) {
   auto const actual_primitive = pb_->Primitive();
 
   // The primitive was computed using Mathematica.
-  auto const expected_primitive = [=](Time const& t){
+  auto const expected_primitive = [=](Time const& t) {
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:264]: Why the change?

File: numerics/poisson_series_test.cpp [hunk 12]
```diff
@@ -377,34 +435,29 @@ TEST_F(PiecewisePoissonSeriesTest, Action) {
     auto const s2 = pp_ + p_;
     EXPECT_THAT(s1(t0_ + 0.5 * Second),
                 AlmostEquals((10 - 3 * Sqrt(2)) / 4, 0));
-    EXPECT_THAT(s1(t0_ + 1.5 * Second),
-                AlmostEquals((6 + Sqrt(2)) / 4, 0));
+    EXPECT_THAT(s1(t0_ + 1.5 * Second), AlmostEquals((6 + Sqrt(2)) / 4, 0));
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:438]: Please revert the changes to lines 438-448.

File: numerics/poisson_series_test.cpp [hunk 13]
```diff
@@ -417,34 +470,29 @@ TEST_F(PiecewisePoissonSeriesTest, ActionMultiorigin) {
     auto const s2 = pp_ + p;
     EXPECT_THAT(s1(t0_ + 0.5 * Second),
                 AlmostEquals((10 - 3 * Sqrt(2)) / 4, 1));
-    EXPECT_THAT(s1(t0_ + 1.5 * Second),
-                AlmostEquals((6 + Sqrt(2)) / 4, 0));
+    EXPECT_THAT(s1(t0_ + 1.5 * Second), AlmostEquals((6 + Sqrt(2)) / 4, 0));
```
Comment (pleroy) [inline numerics/poisson_series_test.cpp:473]: Please revert the changes to lines 473-483.

File: numerics/poisson_series_test.cpp (hunk 11 above)
Comment (eggrobin) [reply numerics/poisson_series_test.cpp:264]: Because it was wrong.

File: numerics/poisson_series_test.cpp (hunk 8 above)
Comment (eggrobin) [reply numerics/poisson_series_test.cpp:240]: Not a chance. It is a solution of some ugly transcendental equation that depends on the phase of the signal.

File: numerics/poisson_series.hpp (hunk 2 above)
Comment (eggrobin) [reply numerics/poisson_series.hpp:313]: > I'd favour giving a name to this type.

meh.

> Also, this file should include `functional`.

Done.
